src/Microsoft.ML.PCA/PcaTransform.cs

```diff
@@ -554,12 +696,122 @@ private static void TransformFeatures(IExceptionContext ectx, ref VBuffer<Float>
         public static CommonOutputs.TransformOutput Calculate(IHostEnvironment env, Arguments input)
         {
             var h = EntryPointUtils.CheckArgsAndCreateHost(env, "Pca", input);
-            var view = new PcaTransform(h, input, input.Data);
+            var view = PcaTransform.Create(h, input, input.Data);
             return new CommonOutputs.TransformOutput()
             {
                 Model = new TransformModel(h, view, input.Data),
                 OutputData = view
             };
         }
     }
+
+    public sealed class PcaEstimator : IEstimator<PcaTransform>
+    {
+        private readonly IHost _host;
+        private readonly PcaTransform.ColumnInfo[] _columns;
+
+        /// <summary>Convinence constructor for simple one column case.</summary>
+        /// <include file='doc.xml' path='doc/members/member[@name="PCA"]/*'/>
+        /// <param name="env">The environment.</param>
+        /// <param name="inputColumn">Input column to apply PCA on.</param>
+        /// <param name="outputColumn">Output column. Null means <paramref name="inputColumn"/> is replaced.</param>
+        /// <param name="weightColumn">The name of the weight column.</param>
+        /// <param name="rank">The number of components in the PCA.</param>
+        /// <param name="overSampling">Oversampling parameter for randomized PCA training.</param>
+        /// <param name="center">If enabled, data is centered to be zero mean.</param>
+        /// <param name="seed">The seed for random number generation</param>
+        public PcaEstimator(IHostEnvironment env, string inputColumn, string outputColumn = null,
+            string weightColumn = PcaTransform.Defaults.WeightColumn, int rank = PcaTransform.Defaults.Rank,
+            int overSampling = PcaTransform.Defaults.Oversampling, bool center = PcaTransform.Defaults.Center,
+            int? seed = null)
+            : this(env, new PcaTransform.ColumnInfo(inputColumn, outputColumn ?? inputColumn, weightColumn, rank, overSampling, center, seed))
+        {
+        }
+
+        public PcaEstimator(IHostEnvironment env, params PcaTransform.ColumnInfo[] columns)
+        {
+            Contracts.CheckValue(env, nameof(env));
+            _host = env.Register(nameof(PcaEstimator));
+            _columns = columns;
+        }
+
+        public PcaTransform Fit(IDataView input) => new PcaTransform(_host, input, _columns);
+
+        public SchemaShape GetOutputSchema(SchemaShape inputSchema)
+        {
+            _host.CheckValue(inputSchema, nameof(inputSchema));
+            var result = inputSchema.Columns.ToDictionary(x => x.Name);
+            foreach (var colInfo in _columns)
+            {
+                if (!inputSchema.TryFindColumn(colInfo.Input, out var col))
+                    throw _host.ExceptSchemaMismatch(nameof(inputSchema), "input", colInfo.Input);
+
+                if (!(col.Kind == SchemaShape.Column.VectorKind.Vector && col.ItemType.IsNumber))
```

Comment:
>col.ItemType.IsNumber [](start = 74, length = 21)

Just want to double check. Do we really support all number types? Doubles, longs?

Reply:
the original check was with .IsNumber, but you're right, it only works with floats. I changed the check.

---
In reply to: [227048656](https://github.com/dotnet/machinelearning/pull/1333#discussion_r227048656) [](ancestors = 227048656)